pkg/dmaasbackup/schedule.go

```diff
@@ -0,0 +1,175 @@
+		if activeSchedule != nil {
+			err = d.scheduleClient.Delete(
+				activeSchedule.ScheduleName,
+				&metav1.DeleteOptions{},
+			)
+			if err != nil {
+				if !apierrors.IsNotFound(err) {
+					d.logger.WithError(err).
+						Errorf("failed to delete schedule=%s", activeSchedule.ScheduleName)
+					return errors.Wrapf(err,
+						"failed to delete schedule=%s", activeSchedule.ScheduleName)
+				}
+			}
+			activeSchedule.Status = v1alpha1.Deleted
+			d.logger.Infof("Schedule=%s deleted", activeSchedule.ScheduleName)
+		}
+
+		// create a new schedule
+		newSchedule, err := d.createSchedule(obj)
+		if err != nil {
+			d.logger.WithError(err).Errorf("failed to create new schedule")
+			return err
+		}
+		appendVeleroSchedule(obj, newSchedule)
+		d.logger.Infof("Schedule=%s created", newSchedule.Name)
+		return nil
+	}
+
+	d.logger.Debugf("Full Backup is not due, next due after %s", nextDue.String())
+	return nil
+}
+
+func getNextDue(cr cron.Schedule, schedule *v1alpha1.VeleroScheduleDetails, now time.Time) (bool, time.Duration) {
+	var lastSync time.Time
+	if schedule != nil && schedule.CreationTimestamp != nil {
+		lastSync = schedule.CreationTimestamp.Time
+	}
+
+	nextSync := cr.Next(lastSync)
+	return now.After(nextSync), nextSync.Sub(lastSync)
+}
+
+func (d *dmaasBackup) createSchedule(dbkp *v1alpha1.DMaaSBackup) (*velerov1api.Schedule, error) {
+	name := d.generateScheduleName(*dbkp)
+	scheduleObj := velerobuilder.ForSchedule(d.velerons, name).
+		Template(dbkp.Spec.VeleroScheduleSpec.Template).
+		CronSchedule(dbkp.Spec.VeleroScheduleSpec.Schedule).
+		Result()
+
+	return d.scheduleClient.Create(scheduleObj)
+}
+
+func (d *dmaasBackup) createBackup(dbkp *v1alpha1.DMaaSBackup) (*velerov1api.Backup, error) {
+	name := d.generateBackupName(*dbkp)
+	backupObj := velerobuilder.ForBackup(d.velerons, name).
+		FromSchedule(&velerov1api.Schedule{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: d.velerons,
+				Name:      name,
+			},
+			Spec: velerov1api.ScheduleSpec{
+				Template: dbkp.Spec.VeleroScheduleSpec.Template,
+			},
+		}).
+		Result()
+
+	return d.backupClient.Create(backupObj)
+}
+
+func (d *dmaasBackup) generateScheduleName(dbkp v1alpha1.DMaaSBackup) string {
+	return dbkp.Name + "-" + d.clock.Now().Format("20060102150405") + "-" + dmaasScheduleSuffix
+}
+
+func (d *dmaasBackup) generateBackupName(dbkp v1alpha1.DMaaSBackup) string {
+	return dbkp.Name + "-" + dmaasScheduleSuffix
+}
+
+func (d *dmaasBackup) processNonFullBackupSchedule(dbkp *v1alpha1.DMaaSBackup) error {
+	if dbkp.Spec.PeriodicFullBackupCfg.CronTime != "" {
+		return nil
+	}
+
+	d.logger.Debug("Processing non fullbackup")
+
+	// non full backup schedule can be of two types, schedule or normal backup
+	// check for scheduled backup
+	if dbkp.Spec.VeleroScheduleSpec.Schedule != "" {
+		// dmaasbackup has schedule spec
+		// create schedule if not created
+		if len(dbkp.Status.VeleroSchedules) == 0 {
+			schedule, err := d.createSchedule(dbkp)
```

Comment:
what happens if Velero Schedule is created successfully, and dmaas-operator got restarted?